### PR TITLE
`EquilibriumAggregation` global aggregation layer

### PR DESCRIPTION
**Pull request by Padarn**
Implements a simple version of the paper ["Equilibrium Aggregation: Encoding Sets via Optimization"](https://arxiv.org/abs/2202.12795). Note that exact details of the paper have not been implemented yet, and I plan to leave specific details about the optimizer out of this MR, it should only add a new readout layer than implements the method.

TODO:
- [x] Example of the 'exact' methods converging (median)
- [x] Cleanup of the readout layer
- [x] Support batched datasets (similar to other glob layers)
- [x] Tests

Addresses https://github.com/pyg-team/pytorch_geometric/issues/4447

### CODE REVIEW

**Comment by Padarn** [comment]
This is still pretty heavy WIP (and I'm not even sure if it is working 100% yet). 

@rusty1s do you think there is a good example graph classification problem to try and compare accuracy to? The paper uses the `MOLPCBA` benchmark (would I could not see in the repo yet, though we could add it). I currently just modified the `MUTAG` example, there is no performance improvement (or degradation).

I plan also have the median example from the paper as an example (acts as a reasonable test it works).

File: torch_geometric/nn/glob/equilibrium_aggregation.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    def reg(self, y):
+        return torch.nn.Softplus()(
+            self.lamb) * (y + EPS).square().mean().sqrt()
+
+    def combine_input(self, x, y):
+        return torch.cat([x, y.expand(x.size(0), -1)], dim=1)
+
+    def forward(self, x: torch.Tensor):
+        grad_enabled = torch.is_grad_enabled()
```
Comment (Padarn) [inline torch_geometric/nn/glob/equilibrium_aggregation.py:60]: I don't know if there is cleaner way to do this - I need to undo other 'torch.no_grad' otherwise the inner loop cannot run in test/validation stages that use it

File: examples/mutag_gin_equib.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+train_loader = DataLoader(train_dataset, batch_size=1)
+test_loader = DataLoader(test_dataset, batch_size=1)
+
+
+class Net(torch.nn.Module):
+    def __init__(self, in_channels, dim, out_channels):
+        super().__init__()
+
+        self.conv1 = GINConv(
```
Comment (rusty1s) [inline examples/mutag_gin_equib.py:25]: Replace with `torch.nn.models.GIN`.

File: examples/mutag_gin_equib.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        self.readout = EquilibriumAggregation(dim, dim, [256, 256],
+                                              grad_iter=10, alpha=0.05)
+
+    def forward(self, x, edge_index, batch):
+        x = self.conv1(x, edge_index)
+        x = self.conv2(x, edge_index)
+        x = self.conv3(x, edge_index)
+        x = self.conv4(x, edge_index)
+        x = self.conv5(x, edge_index)
+        x = self.readout(x).unsqueeze(0)
```
Comment (rusty1s) [inline examples/mutag_gin_equib.py:56]:
```suggestion
        x = self.readout(x)
```
Why do we need the `unsqueeze` here?

File: torch_geometric/nn/glob/__init__.py [hunk 4]
```diff
@@ -3,15 +3,12 @@
...
-    'global_add_pool',
-    'global_mean_pool',
-    'global_max_pool',
-    'global_sort_pool',
-    'GlobalAttention',
-    'Set2Set',
-    'GraphMultisetTransformer',
+    'global_add_pool', 'global_mean_pool', 'global_max_pool',
+    'global_sort_pool', 'GlobalAttention', 'Set2Set',
+    'GraphMultisetTransformer', 'EquilibriumAggregation'
```
Comment (rusty1s) [inline torch_geometric/nn/glob/__init__.py:11]:
```suggestion
    'GraphMultisetTransformer', 'EquilibriumAggregation',
```

File: torch_geometric/nn/glob/equilibrium_aggregation.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+        return h
+
+
+class EquilibriumAggregation(torch.nn.Module):
+    """
+    Args:
+        potential (torch.nn.Module): trainable potenial function
+    """
+    def __init__(self, input_dim: int, output_dim: int, layers: List[int],
```
Comment (rusty1s) [inline torch_geometric/nn/glob/equilibrium_aggregation.py:39]:
```suggestion
    def __init__(self, in_channels: int, out_channels: int, num_layers: List[int],
```
Keep naming consistent :)

File: torch_geometric/nn/glob/equilibrium_aggregation.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+                 grad_iter: int = 5, alpha: float = 0.1):
+        super().__init__()
+
+        self.potential = ResNetPotential(input_dim + output_dim, layers)
+        self.lamb = torch.nn.Parameter(torch.Tensor([1]), requires_grad=True)
```
Comment (rusty1s) [inline torch_geometric/nn/glob/equilibrium_aggregation.py:44]: Looks like we need to fill this tensor in `reset_parameters`.

File: torch_geometric/nn/glob/equilibrium_aggregation.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        self.grad_iter = grad_iter
+        self.alpha = alpha
+        self.output_dim = output_dim
+
+    def init_output(self):
+        return torch.zeros(self.output_dim, requires_grad=True)
+
+    def reg(self, y):
+        return torch.nn.Softplus()(
```
Comment (rusty1s) [inline torch_geometric/nn/glob/equilibrium_aggregation.py:53]: Init `Softplus` in the module constructor?

File: torch_geometric/nn/glob/equilibrium_aggregation.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            self.lamb) * (y + EPS).square().mean().sqrt()
```
Comment (rusty1s) [inline torch_geometric/nn/glob/equilibrium_aggregation.py:54]: Why do we need the `EPS` here?

File: torch_geometric/nn/glob/equilibrium_aggregation.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+    def combine_input(self, x, y):
+        return torch.cat([x, y.expand(x.size(0), -1)], dim=1)
+
+    def forward(self, x: torch.Tensor):
```
Comment (rusty1s) [inline torch_geometric/nn/glob/equilibrium_aggregation.py:59]:
```suggestion
    def forward(self, x: torch.Tensor) -> torch.Tensor:
```

File: torch_geometric/nn/glob/equilibrium_aggregation.py (hunk 1 above)
Comment (rusty1s) [reply torch_geometric/nn/glob/equilibrium_aggregation.py:60]: I think the preferred approach is:
```python
with torch.enable_grad():
    pass
```

File: examples/mutag_gin_equib.py (hunk 2 above)
Comment (Padarn) [reply examples/mutag_gin_equib.py:25]: Yeah sorry this is just lifted from the previous example for now

File: examples/mutag_gin_equib.py (hunk 3 above)
Comment (Padarn) [reply examples/mutag_gin_equib.py:56]: right now the output shape of self.readout(x) is `(input_size)` but needs to be `(batch_size, input_size)` - the disparity comes because the model here is using batch but I've currently ignored that by setting batch to 1 because I wanted to simplify it as a I figured it out

File: torch_geometric/nn/glob/equilibrium_aggregation.py (hunk 8 above)
Comment (Padarn) [reply torch_geometric/nn/glob/equilibrium_aggregation.py:54]: because I didn't realize I should be using `.norm` 😓 https://github.com/pytorch/pytorch/issues/6394

**Comment by Padarn** [comment]
Hey @rusty1s this is mostly ready for review (though the example I use is of the median aggregation).

I'm currently trying to play around to get this to converge well, but I haven't had much luck. I'm not sure if its because of parameters I've used or details missing in the paper (like the alpha used in the inner optimization). But perhaps convergence is just slow - they run for 10 million steps which I'm trying to do now but will take quite some time on my resources :-)

**Comment by Padarn** [comment]
after ~1 million iterations no signs of convergence, guessing I've got something wrong

File: torch_geometric/nn/glob/equilibrium_aggregation.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        batch: Optional[torch.Tensor],
+        func: Callable[[torch.Tensor, torch.Tensor, Optional[torch.Tensor]],
+                       torch.Tensor],
+        iterations: int = 5,
+    ) -> Tuple[torch.Tensor, float]:
+
+        momentum = torch.zeros_like(y)
+        for _ in range(iterations):
+            val = func(x, y, batch)
+            grad = torch.autograd.grad(val, y, create_graph=True,
```
Comment (Padarn) [inline torch_geometric/nn/glob/equilibrium_aggregation.py:96]: The authors of the paper suggest adding an axillary loss on the `grad` here (`grad.square().sum()/iterations`) to be precise.

**Comment by Padarn** [comment]
I've moved this to the aggregation module but there are a few things missing which I'll need opinions on. Leaving comments in the code.

File: torch_geometric/nn/aggr/equilibrium.py [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+        return torch.zeros(index_size, self.output_dim,
+                           requires_grad=True).float()
+
+    def reg(self, y: Tensor) -> float:
+        return self.lamb * y.square().mean(dim=-2).sum(dim=0)
+
+    def energy(self, x: Tensor, y: Tensor, index: Optional[Tensor]):
+        return self.potential(x, y, index) + self.reg(y)
+
+    def forward(self, x: Tensor, index: Optional[Tensor] = None, *,
```
Comment (Padarn) [inline torch_geometric/nn/aggr/equilibrium.py:164]: Just a note this was all written before the `Aggregation` module, so its not using `reduce` yet - need to rethink some of the implementation to make use of it.

File: torch_geometric/nn/aggr/equilibrium.py [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+                ptr: Optional[Tensor] = None, dim_size: Optional[int] = None,
+                dim: int = -2) -> Tensor:
+
+        if ptr is not None:
+            raise ValueError(f"{self.__class__} doesn't support `ptr`")
+
+        if dim_size is not None:
```
Comment (Padarn) [inline torch_geometric/nn/aggr/equilibrium.py:171]: I could technically support this, but I don't know what you'd expect the behavior to be: Unlike sum/mean etc, we can't just assume the input is zero (or if we do, the output would just be random).

Thoughts?

File: torch_geometric/nn/aggr/equilibrium.py (hunk 11 above)
Comment (rusty1s) [reply torch_geometric/nn/aggr/equilibrium.py:164]: No need to make use of `reduce`. The `LSTMAggregation` module doesn't make use of it as well.

File: torch_geometric/nn/aggr/equilibrium.py (hunk 12 above)
Comment (rusty1s) [reply torch_geometric/nn/aggr/equilibrium.py:171]: Please have a look at https://pytorch-scatter.readthedocs.io/en/latest/functions/scatter.html. The `dim_size` argument ideally is identical to `index_size = index.max() + 1` in your case. If passed, there is no need to compute it in the first place.

Comment (Padarn) [reply torch_geometric/nn/aggr/equilibrium.py:171]: I didn't fully get that sorry. If `dim_size` is passed 
- if `dim_size < index_size` I'd expect error,
- if  `dim_size > index_size` I'm not sure what to expect - I guess zero in for those entries?

Comment (rusty1s) [reply torch_geometric/nn/aggr/equilibrium.py:171]: Yes, that is correct.

File: examples/equilibrium_median.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+gamma = torch.distributions.gamma.Gamma(0.2, 0.5)
+uniform = torch.distributions.uniform.Uniform(0, 1)
+total_loss = 0
+n_loss = 0
+for i in range(steps):
+    optimizer.zero_grad()
+    dist = np.random.choice([norm, gamma, uniform])
+    x = dist.sample((input_size, 1))
+    y = model(x)
+    loss = (y - x.median()).norm(2)
```
Comment (lightaime) [inline examples/equilibrium_median.py:31]: It seems the loss is not normalized by the input size.
```suggestion
    loss = (y - x.median()).norm(2) / input_size
```

File: torch_geometric/nn/aggr/equilibrium.py [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+    def forward(
+        self,
+        x: Tensor,
+        y: Tensor,
+        index: Optional[Tensor],
+        func: Callable[[Tensor, Tensor, Optional[Tensor]], Tensor],
+        iterations: int = 5,
+    ) -> Tuple[Tensor, float]:
+
+        momentum = torch.zeros_like(y)
```
Comment (lightaime) [inline torch_geometric/nn/aggr/equilibrium.py:93]:
```suggestion
        momentum_buffer = torch.zeros_like(y)
```

File: torch_geometric/nn/aggr/equilibrium.py [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+        for _ in range(iterations):
+            val = func(x, y, index)
+            grad = torch.autograd.grad(val, y, create_graph=True,
+                                       retain_graph=True)[0]
+            momentum = self.momentum * momentum - self.learning_rate * grad
```
Comment (lightaime) [inline torch_geometric/nn/aggr/equilibrium.py:98]:
```suggestion
            momentum_buffer = self.momentum * momentum_buffer - self.learning_rate * grad
```

**Review by lightaime — COMMENTED** [review]
Great work! Added some minor comments.

**Comment by Padarn** [comment]
thanks @lightaime - I've got a busy week, I'll address your comments ASAP :-)

**Comment by Padarn** [comment]
Thanks for the reviews - I've updated based on most comments. I still have an uncertainty here: https://github.com/pyg-team/pytorch_geometric/pull/4522#discussion_r872945963 but it maybe for another PR

**Comment by rusty1s** [comment]
Thanks @Padarn. Can you also resolve the merge conflicts? I will take a look soon, and address the unresolved comment as well.

**Comment by Padarn** [comment]
Hey @rusty1s, any thoughts on this one? Happy to keep working on it or break it up, but might be good to not leave it here unfinished.

**Comment by rusty1s** [comment]
Yes, no that #4779 is merged, let's try to integrate this next :)

**Comment by Padarn** [comment]
Cool. I'll make another pass to see if there is anything I can clean up based on https://github.com/pyg-team/pytorch_geometric/pull/4779

**Review by lightaime — APPROVED** [review]
Thanks, LGTM!

**Comment by Padarn** [comment]
Thanks @lightaime - I'm going to merge this one, but will try and build some more example use cases before we release the new version so make sure its good.
